PROTOCOL.md

```diff
@@ -145,30 +146,71 @@ Content-Length: 27
+Twirp errors are a JSON object with the keys:
 
 * **code**: One of the Twirp error codes as a string.
 * **msg**: A human-readable message describing the error
   as a string.
-* **meta**: An object with string keys and values holding
+* **meta**: (optional) An object with string values holding
   arbitrary additional metadata describing the error.
 
 Example:
+
+```json
+{
+  "code": "internal",
+  "msg": "Something went wrong"
+}
 ```
+
+Example with metadata:
+
+```json
 {
-    "code": "permission_denied",
-    "msg": "thou shall not pass",
-    "meta": {
-        "target": "Balrog"
-    }
+  "code": "permission_denied",
+  "msg": "Thou shall not pass",
+  "meta": {
+    "target": "Balrog",
+    "power": "999"
+  }
 }
 ```
 
-For more information, see https://github.com/twitchtv/twirp/wiki/Errors.
+### Error Codes
+
+Twirp errors always include an error code. This code is represented
+as a string and must be one of a fixed set of codes, listed in the
+table below. Each code has an associated HTTP Status Code. When a
+server responds with the given error code, it must set the
+corresponding HTTP Status Code for the response.
+
+| Twirp Error Code    | HTTP Status | Description
+| ------------------- | ----------- | -----------
+| canceled            | 408 | The operation was cancelled.
+| unknown             | 500 | An unknown error occurred. For example, this can be used when handling errors raised by APIs that do not return any error information.
+| invalid_argument    | 400 | The client specified an invalid argument. This indicates arguments that are invalid regardless of the state of the system (i.e. a malformed file name, required argument, number out of range, etc.).
+| malformed           | 400 | The client sent a message which could not be decoded. This may mean that the message was encoded improperly or that the client and server have incompatible message definitions.
+| deadline_exceeded   | 408 | Operation expired before completion. For operations that change the state of the system, this error may be returned even if the operation has completed successfully (timeout).
+| not_found           | 404 | Some requested entity was not found.
+| bad_route           | 404 | The requested URL path wasn't routable to a Twirp service and method. This is returned by generated server code and should not be returned by application code (use "not_found" or "unimplemented" instead).
+| already_exists      | 409 | An attempt to create an entity failed because one already exists.
+| permission_denied   | 403 | The caller does not have permission to execute the specified operation. It must not be used if the caller cannot be identified (use "unauthenticated" instead).
+| unauthenticated     | 401 | The request does not have valid authentication credentials for the operation.
+| resource_exhausted  | 403 | Some resource has been exhausted, perhaps a per-user quota, or perhaps the entire file system is out of space.
```

Comment:
Does this need to be updated to 429?

Reply:
Yes, this entire file should be a copy-paste form the spec in the docs. Which I though I did already 🤔  good catch. I wonder if this should just contain a link to the docs ... but realistically it almost never changes, so it's fine to have some duplication 🤷

Reply:
oh.. I see. I need to merge #270 first